---
title: "Real data analysis - colorectal cancer Visium data"
author: "Zijian Ni"
date: "`r Sys.Date()`"
output: html_document
---

```{r, echo=F}
library(tidyverse)
library(cowplot)
library(SpotClean)
library(Seurat)
library(Matrix)
library(reshape2)
library(SPOTlight)
library(RColorBrewer)
library(BayesSpace)


SEED <- 2021

myPalette <- colorRampPalette(rev(RColorBrewer::brewer.pal(11, "Spectral")))
col_normal <- "#F7b6cc"
col_tumor <- "#0e1058"
tif_dir <- "~/Google Drive/Hallu/codes/ckgroup/spatial_data/manuscript/figures/Rplots"

# paths to Visium raw data
spatial_dir <- "~/Google Drive/Hallu/DATA/10x/Spatial/Parent_Visium_Human_ColorectalCancer/Parent_Visium_Human_ColorectalCancer_raw_feature_bc_matrix.h5"
spatial_slide_dir <- "~/Google Drive/Hallu/DATA/10x/Spatial/Parent_Visium_Human_ColorectalCancer/spatial/tissue_positions_list.csv"
spatial_img_dir <- "~/Google Drive/Hallu/DATA/10x/Spatial/Parent_Visium_Human_ColorectalCancer/spatial/tissue_lowres_image.png"
scale_factor_dir <- "~/Google Drive/Hallu/DATA/10x/Spatial/Parent_Visium_Human_ColorectalCancer/spatial/scalefactors_json.json"

tumor_label_dir <- "~/Google Drive/Hallu/DATA/10x/Spatial/Parent_Visium_Human_ColorectalCancer/is_tumor3.csv"

# paths to scRNAseq data
sc_normal_dir <- "~/Google Drive/Hallu/DATA/Fluidigm/colorectal_tumor_atlas/GSE81861_CRC_NM_all_cells_COUNT.csv/GSE81861_CRC_NM_all_cells_COUNT.csv"
sc_tumor_dir <- "~/Google Drive/Hallu/DATA/Fluidigm/colorectal_tumor_atlas/GSE81861_CRC_tumor_all_cells_COUNT.csv/GSE81861_CRC_tumor_all_cells_COUNT.csv"

```

### Process reference single-cell data

```{r, echo=F, eval=F}
# load single-cell reference data
clean_sc <- function(sc_dir, cutoff=1){
    sc_dat <- read_csv(sc_dir)
    row_names <- sc_dat$X1
    sc_dat <- select(sc_dat, -X1)
    cell_ids <- gsub("__.*","",colnames(sc_dat))
    cell_types <- gsub(".*__(.+)__.*","\\1",colnames(sc_dat))
    sc_dat <- as.matrix(sc_dat) %>% as("dgCMatrix")
    rownames(sc_dat) <- row_names
    colnames(sc_dat) <- cell_ids
    cell_types <- data.frame(cell_id=cell_ids, cell_type=cell_types)
    sc_dat <- sc_dat[rowMeans(sc_dat)>=cutoff,]
    return(list(sc_dat, cell_types))
}

sc_normal_list <- clean_sc(sc_normal_dir)
sc_tumor_list <- clean_sc(sc_tumor_dir)

cell_types <- rbind(mutate(sc_normal_list[[2]], maignancy="normal"), 
                    mutate(sc_tumor_list[[2]], maignancy="tumor"))

com_gene <- intersect(rownames(sc_normal_list[[1]]), rownames(sc_tumor_list[[1]]))

sc_dat <- cbind(sc_normal_list[[1]][com_gene,], sc_tumor_list[[1]][com_gene,])
row_names <- gsub(".*_(.+)_.*","\\1",rownames(sc_dat))
sc_dat <- sc_dat[!duplicated(row_names),] # remove duplicated genes
rownames(sc_dat) <- row_names[!duplicated(row_names)]


save(sc_dat, cell_types, file="~/Google Drive/Hallu/DATA/Fluidigm/colorectal_tumor_atlas/scRNAseq_preprocess.RData")
```

Dimension reduction and visualization.

```{r, echo=F, eval=F}
load("~/Google Drive/Hallu/DATA/Fluidigm/colorectal_tumor_atlas/scRNAseq_preprocess.RData")

set.seed(SEED)
S_sc <- CreateSeuratObject(sc_dat) %>% 
    SCTransform(verbose=F) %>%
    RunPCA(verbose = FALSE) %>%
    RunUMAP(dims = 1:30, verbose = FALSE)

S_sc$malignancy <- cell_types$maignancy
S_sc$celltype <- cell_types$cell_type

DimPlot(S_sc,
        group.by = "celltype",
        label = TRUE) + NoLegend()
DimPlot(S_sc,
        group.by = "malignancy",
        label = TRUE) + NoLegend()

# DE marker genes
Idents(S_sc) <- "malignancy"
cluster_markers_all <- FindAllMarkers(object = S_sc,
                                      slot = "data",
                                      verbose = F, 
                                      only.pos = TRUE)

DE_genes <- filter(cluster_markers_all, p_val_adj<=0.01)$gene
```

### Process Visium data

```{r, echo=F, eval=F}
# load Visium data
bc_raw <- Read10xRawH5(spatial_dir)
bc_slide <- Read10xSlide(spatial_slide_dir,spatial_img_dir, scale_factor_dir)
slide_obj <- CreateSlide(bc_raw, bc_slide)

set.seed(SEED)
decont_obj <- SpotClean(slide_obj)

save(slide_obj, decont_obj, file="~/Google Drive/Hallu/DATA/10x/Spatial/Parent_Visium_Human_ColorectalCancer/raw_decont.RData")
```

### Check tumor labels

```{r, echo=F, dpi=300, fig.width=6, fig.height=5}
load("~/Google Drive/Hallu/DATA/10x/Spatial/Parent_Visium_Human_ColorectalCancer/raw_decont.RData")
tumor_labels <- read_csv(tumor_label_dir)
decont_obj@metadata$slide <- inner_join(decont_obj@metadata$slide, tumor_labels, by=c("barcode"="Barcode"))

dp <- VisualizeLabel(decont_obj,"is_tumor")+
    scale_fill_manual(values = c(col_normal,col_tumor),breaks = c("normal","tumor"))+
    theme_nothing()

# tiff(file.path(tif_dir,paste0("colorectal_cancer_visual_label.tif")),
#      width = 6.8, height = 6,units = "in",res = 300)
plot(dp)
# dev.off()
```

### Decompose spatial data

Decompose raw data and decontaminated data respectively.

```{r, echo=F, eval=F}
set.seed(SEED)

spotlight_raw <- spotlight_deconvolution(
    se_sc = S_sc,
    counts_spatial = slide_obj@assays@data$raw[decont_obj@metadata$decontaminated_genes,colnames(decont_obj)],
    clust_vr = "malignancy", cluster_markers = cluster_markers_all, cl_n = 100, 
    hvg = 3000, ntop = NULL, transf = "uv", method = "nsNMF", min_cont = 0 
)

set.seed(SEED)

spotlight_decont <- spotlight_deconvolution(
    se_sc = S_sc,
    counts_spatial = decont_obj@assays@data$decont[decont_obj@metadata$decontaminated_genes,],
    clust_vr = "malignancy", cluster_markers = cluster_markers_all, cl_n = 100,
    hvg = 3000, ntop = NULL, transf = "uv", method = "nsNMF", min_cont = 0 
)


# save all the above results for later analyses
save(spotlight_raw, spotlight_decont, cluster_markers_all, file = "~/Google Drive/Hallu/DATA/10x/Spatial/Parent_Visium_Human_ColorectalCancer/spotlight_decomp.RData")

```

### Load and clean up decomposed data

```{r, echo=F, message=F, warning=F}
load("~/Google Drive/Hallu/DATA/10x/Spatial/Parent_Visium_Human_ColorectalCancer/spotlight_decomp.RData")

# prepare for SPOTlight inputs
nmf_mod_list <- list(raw=spotlight_raw[[1]], SpotClean=spotlight_decont[[1]])
decon_mtrx_list <- list(raw=spotlight_raw[[2]], SpotClean=spotlight_decont[[2]])

S_sp <- Load10X_Spatial("~/Google Drive/Hallu/DATA/10x/Spatial/Parent_Visium_Human_ColorectalCancer","Parent_Visium_Human_ColorectalCancer_raw_feature_bc_matrix.h5")

decont_mat <- decont_obj@assays@data$decont
rownames(decont_mat) <- gsub("_","-",rownames(decont_mat))
S_sp_list <- list(raw=S_sp[rownames(decont_obj),colnames(decont_obj)], 
                      SpotClean=S_sp[rownames(decont_mat),colnames(decont_mat)])
S_sp_list$SpotClean@assays$Spatial@counts <- 
    S_sp_list$SpotClean@assays$Spatial@data <- decont_mat

for(dset in names(nmf_mod_list)){
    
    rownames(decon_mtrx_list[[dset]]) <- colnames(S_sp_list[[dset]])
    
    decon_df <- decon_mtrx_list[[dset]] %>%
        data.frame() %>%
        tibble::rownames_to_column("barcodes")
    
    S_sp_list[[dset]]@meta.data <- S_sp_list[[dset]]@meta.data %>%
        tibble::rownames_to_column("barcodes") %>%
        dplyr::left_join(decon_df, by = "barcodes") %>%
        tibble::column_to_rownames("barcodes")
    
    S_sp_list[[dset]]$visual_malignant_status <-
        tumor_labels$is_tumor[
            match(colnames(S_sp_list[[dset]]),tumor_labels$Barcode)
        ]
    
    S_sp_list[[dset]]$tumor_comp <-
        decon_mtrx_list[[dset]][colnames(S_sp_list[[dset]]),"tumor"]

}
```


##### Plot estimated malignant composition

```{r, echo=F, dpi=300, fig.width=10, fig.height=5}

# tiff(file.path(tif_dir,"colorectal_cancer_decomp_crop.tif"),res = 300,
#      width = 8, height = 4, units = "in")
plot_grid(
    SpatialFeaturePlot(S_sp_list$raw,features = "tumor_comp",alpha = c(0.1,1),crop = T,
                       min.cutoff = 0.3, max.cutoff = 0.55, pt.size.factor = 2, stroke = 1e-10)+
        NoLegend()+
        labs(fill="Malignant\ncomposition  "),

    SpatialFeaturePlot(S_sp_list$SpotClean,features = "tumor_comp",alpha = c(0.1,1),crop = T,
                       min.cutoff = 0.3, max.cutoff = 0.55, pt.size.factor = 2, stroke = 1e-10)+
        NoLegend()+
        labs(fill="Malignant\ncomposition  "),
    # SpatialDimPlot(S_sp_list$raw[,S_sp_list$raw$visual_malignant_status=="tumor"],
    #                crop = F,pt.size.factor = 1.5, stroke = 1e-10,
    #                group.by = "visual_malignant_status",cols = col_tumor)+NoLegend(),
    ncol=2, align = "vh")
# dev.off()

```


### BayesSpace clustering of tumor spots

```{r, echo=F, eval=F}
bs_list <- list()
set.seed(SEED)
for(dset in names(S_sp_list)){
    tumor_bc <- colnames(S_sp_list[[dset]])[S_sp_list[[dset]]$visual_malignant_status=="tumor"]
    tumor_slide <- filter(decont_obj@metadata$slide, barcode%in%tumor_bc)
    tumor_mat <- S_sp_list[[dset]]@assays$Spatial@counts[,tumor_slide$barcode]
    
    bs_list[[dset]] <- SingleCellExperiment(assays=list(counts=tumor_mat),
                                colData=tumor_slide) %>% 
        spatialPreprocess(platform="Visium") %>% 
        spatialCluster(q=6, platform="Visium", nrep = 10000)
}

save(bs_list, file="~/Google Drive/Hallu/DATA/10x/Spatial/Parent_Visium_Human_ColorectalCancer/BayesSpace_clust_6_tumor_only.RData")

```

### Annotate tumor clusters

```{r, echo=F}
load("~/Google Drive/Hallu/DATA/10x/Spatial/Parent_Visium_Human_ColorectalCancer/BayesSpace_clust_6_tumor_only.RData")
plot_grid(
clusterPlot(bs_list$raw)+labs(title="raw"),
clusterPlot(bs_list$SpotClean)+labs(title="SpotClean"),
ncol=2, align = "vh")

tumor_labels <- read_csv(tumor_label_dir) %>% 
    full_join(as.data.frame(bs_list$raw@colData@listData) %>% 
                   select(barcode, spatial.cluster) %>% dplyr::rename(raw_clust="spatial.cluster"), 
               by=c("Barcode"="barcode")) %>% 
    full_join(as.data.frame(bs_list$SpotClean@colData@listData) %>% 
                   select(barcode, spatial.cluster) %>% dplyr::rename(SpotClean_clust="spatial.cluster"), 
               by=c("Barcode"="barcode")) %>%
    mutate(raw_clust=ifelse(is.na(raw_clust),"non-tumor",paste0("tumor_",raw_clust)), 
           SpotClean_clust=ifelse(is.na(SpotClean_clust),"non-tumor",paste0("tumor_",SpotClean_clust)))

# make them lined up for better visualization
tumor_labels <- mutate(
    tumor_labels,
    raw_clust=ifelse(
        raw_clust=="non-tumor", "non-tumor", ifelse(
            raw_clust=="tumor_3","tumor_4",ifelse(
                raw_clust=="tumor_4", "tumor_5", ifelse(
                    raw_clust=="tumor_5", "tumor_6", ifelse(
                        raw_clust=="tumor_6", "tumor_2", ifelse(
                            raw_clust=="tumor_2", "tumor_3", "tumor_1"
                        )))))))

tumor_labels$tumor_comp_raw <- S_sp_list$raw$tumor_comp[tumor_labels$Barcode]
tumor_labels$tumor_comp_SpotClean <- S_sp_list$SpotClean$tumor_comp[tumor_labels$Barcode]

```


### Show tumor clusters in spatial slide

```{r, echo=F}
decont_obj@metadata$slide$raw_clust <- NULL
decont_obj@metadata$slide$SpotClean_clust <- NULL

decont_obj@metadata$slide <- inner_join(decont_obj@metadata$slide %>% select(-is_tumor), 
                                        tumor_labels, by=c("barcode"="Barcode"))

# better color
cols <- RColorBrewer::brewer.pal(8,"Set1")
cols[6] <- "#d4d415"
cols[7] <- "#66c2ff"

# tiff(file.path(tif_dir,"colorectal_cancer_spatial_clusters.tif"),res = 300,
#      width = 6, height = 12, units = "in")
plot_grid(
VisualizeLabel(decont_obj,"raw_clust")+
    scale_fill_manual(labels=c("non-tumor",paste0("tumor_",seq_along(unique(tumor_labels$raw_clust)))),
                      values=c("#ebebeb",cols[c(8,2:6)]))+
    labs(fill="") + theme_void()+theme(legend.position = "bottom"),
VisualizeLabel(decont_obj,"SpotClean_clust")+
    scale_fill_manual(labels=c("non-tumor",paste0("tumor_",seq_along(unique(tumor_labels$SpotClean_clust)))),
                      values=c("#ebebeb",cols[c(7,1,3:6)]))+
    labs(fill="") + theme_void()+theme(legend.position = "bottom"),
ncol = 1, align="vh")
# dev.off()

```

### Malignancy compositions in tumor clusters

```{r, echo=F}
malig_score_df <- tumor_labels %>% group_by(SpotClean_clust) %>% 
    summarize(avg_score_raw=mean(tumor_comp_raw), 
              avg_score_SpotClean=mean(tumor_comp_SpotClean))
```


### Differences among tumor clusters

```{r, echo=F}
S_sp_list$raw$tumor_clust <- decont_obj@metadata$slide$raw_clust
S_sp_list$SpotClean$tumor_clust <- decont_obj@metadata$slide$SpotClean_clust

DE_list <- list()
set.seed(SEED)
for(dset in names(S_sp_list)){
    S_sp_list[[dset]] <- S_sp_list[[dset]][,S_sp_list[[dset]]$tumor_clust!="normal"] %>%
    # S_sp_list[[dset]] <- S_sp_list[[dset]] %>%
    NormalizeData(verbose=F)
    Idents(S_sp_list[[dset]]) <- "tumor_clust"
    DE_list[[dset]] <- FindAllMarkers(S_sp_list[[dset]])
}

filter(DE_list$SpotClean, cluster=="tumor_1", p_val_adj<=0.01) %>% str

# 9 IG genes among the top 10 DE genes in cluster 1
IG_genes <- c("IGHG4","IGHG3",
"IGHG1","IGLC1",
"IGKC","IGLC2",
"IGHA1","IGLC3","IGHM")

# tiff(file.path(tif_dir,"colorectal_cancer_IG_markers.tif"),res = 300,
#      width = 9, height = 9, units = "in")
SpatialFeaturePlot(S_sp_list$SpotClean, slot = "data", features = IG_genes, 
                   pt.size.factor = 2, stroke = 0)
# dev.off()
```

